### PR TITLE
Workaround a deadlock caused by watching .editorconfigs

### PR DESCRIPTION
**Pull request by jasonmalinowski**
We use the CodingConventions library from Visual Studio that parses and processes .editorconfig files for us. This library under the covers uses the Visual Studio file watching service for watching files. In Visual Studio 2017 the file watching service made the guarantee that watching files could be done in a free-threaded manner in the background and wouldn't marshal. The .editorconfig library then passed this guarantee along to us.

In Dev16, the file watching service was rewritten, and accidentally lost that guarantee in specific cases. Much of our interactions with the file change service also changed so it doesn't matter, but the .editorconfig library's use of the file watching service is causing deadlocks in some scenarios. This change is a workaround to ensure we don't start watching .editorconfig files on the background thread in places where we need to ensure we don't touch the UI thread.

Ideally we'd be fixing the file change service itself to restore the guarantee, but that's too risky of a change for our current needs and isn't something we can do quickly enough to avoid customer issues. Since we're rewriting our .editorconfig handling entirely which will cause us to stop using the .editorconfig library (and by extension, it's use of this problematic file change service path), we can do this workaround now and let the problem go away on it's own soon when we simply delete all of this.

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/810494

<details><summary>Ask Mode template</summary>

### Customer scenario

User tries to create a new IntelliTest into an existing project. The operation may hang.

### Bugs this fixes

https://devdiv.visualstudio.com/DevDiv/_workitems/edit/810494

### Workarounds, if any

No practical workarounds.

### Risk

Low to moderate. This is a tactical fix designed to limit risk.

### Performance impact

Lowish: we're potentially duplicating the cache of .editorconfig data being used since we're not sharing the same instance as the rest of VS. There is a possibility that an unknown configuration could encounter observable slowdowns, though these are not as risky as deadlocks. We are working with the platform team to address the API change that led to this problem so the workaround, along with its risks, can be removed in a timely manner.

### Is this a regression from a previous update?

Yes, regressed at some point during Dev16.

### Root cause analysis

In Dev16, the file watching service was rewritten, and accidentally lost a free-threading guarantee in specific cases, which causes this deadlock. Rearchitecting it to avoid that assumption is too complicated and risky, so we're doing a fix here.

### How was the bug found?

Internal testing.

</details>

### CODE REVIEW

**Comment by sharwell** [comment]
@jasonmalinowski I have a lot of faith in the ApplyEditorConfigAndFormatDocument integration test, so what I'll be looking for is the reason why it started to fail here and what it takes to correct it.

**Comment by jasonmalinowski** [comment]
@sharwell I guess one likely explanation would be that since I'm now doing things async, there's a window there were I need to stick a waiter in. I'll take a look at it.

**Comment by jasonmalinowski** [comment]
@sharwell: adding a waiter and correctly reporting the change type seems to have stabilized it on my local runs. Change is pushed and we'll see what the CI run says.

**Comment by jasonmalinowski** [comment]
@sharwell @jinuj @dotnet/roslyn-ide: doing overnight test runs is making this look to be quite stable -- I hit an unrelated ancient bug in Roslyn before seeing a failure locally. Thus this is ready for review.

**Comment by sharwell** [comment]
I think the **Performance impact** section of the ask mode template should be updated. Perhaps something like this:

> User-observable impact is not expected, especially in default installation configurations, but the coding conventions library is known to have substantial cost in cases where it is heavily used. There is a possibility that an unknown configuration could encounter observable slowdowns, though these are not as risky as deadlocks. We are working with the platform team to address the API change that led to this problem so the workaround, along with its risks, can be removed in a timely manner.

File: src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProviderFactory.cs [hunk 1]
```diff
@@ -2,32 +2,162 @@
...
+                codingConventionsManager = _codingConventionsManager;
+            }
+            else
+            {
+                // The default file watcher implementation inside Visual Studio accientally depends on the UI thread
+                // (sometimes!) when trying to add a watch to a file. This can cause us to deadlock, since our assumption is
+                // consumption of a coding convention can be done freely without having to use a JTF-friendly wait.
+                // So we'll wrap the standard file watcher with one that defers the file watches until later.
+                var deferredFileWatcher = new DeferredFileWatcher(_fileWatcher, _asynchronousOperationListenerProvider);
+                codingConventionsManager = CodingConventionsManagerFactory.CreateCodingConventionsManager(deferredFileWatcher);
```
Comment (olegtk) [inline src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProviderFactory.cs]: oh, good thing this API is still there

**Pull request merged**